pkg/instancetype/instancetype.go

```diff
@@ -1534,28 +1534,36 @@ func applyFirmwarePreferences(preferenceSpec *instancetypev1beta1.VirtualMachine
 		return
 	}
 
+	firmware := preferenceSpec.Firmware
+
 	if vmiSpec.Domain.Firmware == nil {
 		vmiSpec.Domain.Firmware = &virtv1.Firmware{}
 	}
 
+	vmiFirmware := vmiSpec.Domain.Firmware
+
 	if vmiSpec.Domain.Firmware.Bootloader == nil {
-		vmiSpec.Domain.Firmware.Bootloader = &virtv1.Bootloader{}
+		vmiFirmware.Bootloader = &virtv1.Bootloader{}
+	}
+
+	if firmware.PreferredUseBios != nil && *firmware.PreferredUseBios && vmiFirmware.Bootloader.BIOS == nil && vmiFirmware.Bootloader.EFI == nil {
+		vmiFirmware.Bootloader.BIOS = &virtv1.BIOS{}
 	}
 
-	if preferenceSpec.Firmware.PreferredUseBios != nil && *preferenceSpec.Firmware.PreferredUseBios && vmiSpec.Domain.Firmware.Bootloader.BIOS == nil && vmiSpec.Domain.Firmware.Bootloader.EFI == nil {
-		vmiSpec.Domain.Firmware.Bootloader.BIOS = &virtv1.BIOS{}
+	if firmware.PreferredUseBiosSerial != nil && vmiFirmware.Bootloader.BIOS != nil && vmiFirmware.Bootloader.BIOS.UseSerial == nil {
+		vmiFirmware.Bootloader.BIOS.UseSerial = ptr.To(*firmware.PreferredUseBiosSerial)
 	}
 
-	if preferenceSpec.Firmware.PreferredUseBiosSerial != nil && vmiSpec.Domain.Firmware.Bootloader.BIOS != nil && vmiSpec.Domain.Firmware.Bootloader.BIOS.UseSerial == nil {
-		vmiSpec.Domain.Firmware.Bootloader.BIOS.UseSerial = ptr.To(*preferenceSpec.Firmware.PreferredUseBiosSerial)
+	if firmware.PreferredEfi != nil {
+		vmiFirmware.Bootloader.EFI = firmware.PreferredEfi.DeepCopy()
 	}
 
-	if preferenceSpec.Firmware.PreferredUseEfi != nil && *preferenceSpec.Firmware.PreferredUseEfi && vmiSpec.Domain.Firmware.Bootloader.EFI == nil && vmiSpec.Domain.Firmware.Bootloader.BIOS == nil {
-		vmiSpec.Domain.Firmware.Bootloader.EFI = &virtv1.EFI{}
+	if firmware.DeprecatedPreferredUseEfi != nil && *firmware.DeprecatedPreferredUseEfi && vmiFirmware.Bootloader.EFI == nil && vmiFirmware.Bootloader.BIOS == nil {
```

Comment:
Should we make the old and new fields mutually exclusive? Or output a warning in the admitter?

Reply:
I think the current behaviour is enough, ignoring the use of the deprecated fields if the newer field is used by checking if `vmiFirmware.Bootloader.EFI == nil` here.

Reply:
Ah wait, PreferredUseSecureBoot isn't using that conditional. Let me rework this slightly.

Reply:
Added the following to just return early when using `PreferredEfi`:

https://github.com/kubevirt/kubevirt/pull/12739/commits/3a31137abe464d1560e9b693e91d311273ec69cd#diff-5e2591267ff7e9303f97ccac213986c2157131e3487a04a122b7662d401f7a57R1559